libairspy/src/iqconverter_float.c

```diff
@@ -150,6 +171,21 @@ static _inline float process_fir_taps(const float *kernel, const float *queue, i
 
 			queue += 8;
 			kernel += 8;
+
+#elif defined(USE_NEON)
+
+		for (i = 0; i < it; i++)
+		{
+			const float32x4_t head1 = vld1q_f32(queue);
+			const float32x4_t kern1 = vld1q_f32(kernel);
+			const float32x4_t head2 = vld1q_f32(queue + 4);
+			const float32x4_t kern2 = vld1q_f32(kernel + 4);
+
+			acc = vmlaq_f32(acc, kern1, head1);
+			acc = vmlaq_f32(acc, kern2, head2);
```

Comment:
Actually I think I found a performance issue here. If you use the same accumulator for both FMLA operations, then CPU won't be able to parallelise them. Even if all quad registers are parallel. Try setting up 2 accumulators and sum them up after the loop. It should give you ~2x performance boost. 

I did some analysis here for different number of quad registers: https://dernasherbrezon.com/posts/fir-filter-optimization-simd/ 

<img width="746" alt="Screenshot 2023-06-27 at 00 00 03" src="https://github.com/airspy/airspyone_host/assets/1614424/719a1ce7-bec2-409d-a2ab-9ce8da7e94a9">

Reply:
Thanks! Nice to see someone is digging deeper into the patch! :)

Indeed using multiple accumulators will help in this function. I've committed a small change for it. So you might want to give it another whirl!

Btw, the article is pretty cool! There is one thing to clarify though: our time measurements are different. I was measuring the overall processing time which happens in the `consumer_threadproc` when running `./airspy-tools/src/airspy_rx -t 0 -f 50 -r output.wav`, and not the speedup of individual functions. It seemed to be more practical measure, which more closely resembles the amount of freed-up resources for other calculations.

The code I used for benchmarking is in the `benchmark` branch of my fork: sergeyvfx/airspyone_host@2b6f827f714
One thing to note is that while it seemed to work fine on Raspberry Pi back then, now I was unable to get reliable number on Apple M2. There might be some stupid mistake in the timing code.

Reply:
```iqconverter_float_process``` is the bottleneck. You can run ```Instruments``` application from Xcode and connect to the running airspy_rx.

<img width="851" alt="Screenshot 2023-06-27 at 22 18 51" src="https://github.com/airspy/airspyone_host/assets/1614424/d7ea7651-40d7-45ba-8cd6-3987b614cb76">

FIR filter can be optimised, but one of the most heavily loaded function ```remove_dc``` cannot. It is not SIMD-friendly because on each iteration it uses the result from the previous operation. I tried to figure out how to algorithmically change it, but cannot understand how it removes DC.

Reply:
One of the ways to remove DC from a signal is to remove its average from the signal. That is effectively what is happening in the `remove_dc`. It calculates the average by some sort of simplified exponential moving average. Typically you'd see a Lerp, but this FMA style of average update works good enough and is cheaper.

It also seems `memmove` has considerable contribution to the timing? Perhaps something like double-bufferred circular buffer will help. Will probably help much more on Raspberry Pi, where memory transfers are not that fast. The tricky part is that such approach will not be usable for SSE path, as it ruins data alignment.